pkg/controller/awsmachine/awsmachine_controller.go

```diff
@@ -204,11 +217,66 @@ func (r *ReconcileAWSMachine) reconcile(ctx context.Context, scope *scope.Machin
 	}
 
 	if err := r.reconcileLBAttachment(scope, instance); err != nil {
-		return errors.Errorf("failed to reconcile LB attachment: %+v", err)
+		return reconcile.Result{}, errors.Errorf("failed to reconcile LB attachment: %+v", err)
+	}
+
+	existingSecurityGroups, err := ec2svc.GetInstanceSecurityGroups(*scope.GetInstanceID())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Ensure that the security groups are correct.
+	_, err = r.ensureSecurityGroups(ec2svc, scope, scope.ProviderMachine.Spec.AdditionalSecurityGroups, existingSecurityGroups)
+	if err != nil {
+		return reconcile.Result{}, errors.Errorf("failed to apply security groups: %+v", err)
+	}
+
+	// Ensure that the tags are correct.
+	_, err = r.ensureTags(ec2svc, scope.ProviderMachine, scope.GetInstanceID(), scope.ProviderMachine.Spec.AdditionalTags)
+	if err != nil {
+		return reconcile.Result{}, errors.Errorf("failed to ensure tags: %+v", err)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileAWSMachine) reconcileDelete(scope *scope.MachineScope, awsMachine *infrav1.AWSMachine) (reconcile.Result, error) {
+	scope.Info("Handling deleted AWSMachine")
+
+	ec2Service := ec2.NewService(scope.Parent)
+
+	instance, err := r.findInstance(scope, ec2Service)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if instance == nil {
+		// The machine was never created or was deleted by some other entity
+		scope.V(3).Info("Unable to locate instance by ID or tags")
+		return reconcile.Result{}, nil
 	}
 
-	// Check if we can perform any in-place updates.
-	return r.update(scope, instance)
+	// Check the instance state. If it's already shutting down or terminated,
+	// do nothing. Otherwise attempt to delete it.
+	// This decision is based on the ec2-instance-lifecycle graph at
+	// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-lifecycle.html
+	switch instance.State {
+	case infrav1.InstanceStateShuttingDown, infrav1.InstanceStateTerminated:
+		scope.Info("Instance is shutting down or already terminated")
+		return reconcile.Result{}, nil
+	default:
+		scope.Info("Terminating instance")
+		if err := ec2Service.TerminateInstance(instance.ID); err != nil {
```

Comment:
Should this use `TerminateAndWait()` to avoid blocking Cluster Deletion if this is the last Machine being deleted?

Reply:
Maybe. It depends on how we handle https://github.com/kubernetes-sigs/cluster-api/issues/1190. If we say a Machine always has a Cluster ownerRef, we don't need to wait because the CAPI cluster deletion logic will wait for the Machine to disappear. If we say that a Machine that belongs to a MachineSet does not always have a Cluster ownerRef, then we need to call TerminateAndWait here.

Reply:
In both cases we would be terminating the instance and Deleting the Machine through removing the finalizer, prior to the backing instance being confirmed terminated.

Reply:
That's true. I can switch it to wait. FWIW, this code was copied from 0.3.